fuel-core-interfaces/src/relayer.rs

```diff
@@ -0,0 +1,141 @@
+use std::collections::HashMap;
+
+use async_trait::async_trait;
+use fuel_storage::Storage;
+use fuel_types::{Address, AssetId, Bytes32, Word};
+use serde::{Deserialize, Serialize};
+use tokio::sync::oneshot;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct DepositCoin {
+    pub owner: Address,
+    pub amount: Word,
+    pub asset_id: AssetId,
+    pub deposited_da_height: u64,
+    pub fuel_block_spend: Option<u64>,
+}
+
+// Database has two main functionalities, ValidatorSet and TokenDeposits.
+// From relayer perspective TokenDeposits are just insert when they get finalized.
+// But for ValidatorSet, It is litle bit different.
+#[async_trait]
+pub trait RelayerDb:
+     Storage<Bytes32, DepositCoin, Error = KvStoreError> // token deposit
+    + Storage<Address, u64,Error = KvStoreError> // validator set
+    + Storage<u64, HashMap<Address, u64>,Error = KvStoreError> //validator set diff
+    + Send
+    + Sync
+{
+    /// deposit token to database. Token deposits are not revertable
+    async fn insert_token_deposit(
+        &mut self,
+        deposit_nonce: Bytes32, // this is ID
+        deposited_da_height: u64, // eth block when deposit is made
+        owner: Address,       // owner
+        asset_id: AssetId,
+        amount: Word,
+    ) {
+        let coin = DepositCoin {
+            owner,
+            amount,
+            asset_id,
+            deposited_da_height,
+            fuel_block_spend: None,
+        };
+        // TODO check what id are we going to use
+        // depends on https://github.com/FuelLabs/fuel-specs/issues/106
+        let _ = Storage::<Bytes32, DepositCoin>::insert(self,&deposit_nonce,&coin);
+    }
+
+    /// Asumption is that validator state is already checked in da side, and we can blidly apply
+    /// changed to db without checking if we have enought stake to reduce.
+    /// What needs to be done is to have validator set state and diff as separate database values.
+    async fn insert_validators_diff(&mut self, da_height: u64, stakes: &HashMap<Address, u64>) {
+        let _ = Storage::<u64,HashMap<Address,u64>>::insert(self, &da_height,stakes);
+    }
+
+    /// get stakes difference between fuel blocks. Return vector of changed (some blocks are not going to have any change)
+    async fn get_validator_diffs(
+            &self,
+            from_fuel_block: u64,
+            to_fuel_block: Option<u64>,
```

Comment:
given that we insert diffs using da_height, would it be more accurate to call these `from_da_block` and `to_da_block`?

Reply:
Yeah